[SIL] Add tests for 'undef' values & code fixes

ezephir opened the pull request:
This change validates that 'undef' can appear in most places where values are
expected by the SIL parser. Fixes are also included for the 'select_value'
instruction. This resolves [SR-304](https://bugs.swift.org/browse/SR-304).

ezephir commented:
To make it easy to review these changes, the SIL file is in the same order as the listings in the SIL documentation.

Some questions:
* Is it okay to ignore 80 column wrapping in SIL files?
* What should I do about alloc_stack and dealloc_stack? Undef is not accepted because the validator ensures that the instruction comes in pairs. I believe that the same issue exists when using alloc_ref and dealloc_ref when specifying '[stack].'
* Is it okay to CHECK the entire instruction? On one hand, I know that the instruction is entirely present, but the tests will break on formatting changes.

slavapestov commented:
@jckarter Might want to review this?

jckarter commented:
LGTM.

tkremenek commented:
@ezephir Can you resolve the merge conflicts?

ezephir commented:
Merge conflict fixed and minor updates to make things pass again.

One thing I'm unsure about is that the verifier says that the successful branch of checked_cast_addr_br must be a label that doesn't take any arguments. The SIL documentation still mentions the parameter as being required, but I've removed it to make the verifier happy. If this is the wrong thing to do, please let me know.

gottesmm commented:
@swift-ci Please test

lattner commented:
The patch looks good, and I'd merge it, but it has new conflicts in it. @ezephir, I'm sorry to ask this again, but can you please update the patch?

ezephir commented:
@lattner Thanks for taking a look. I've updated and rebased the PR.

The test cases now fail because they insist that the SIL system parse its own output and the code for "mark_function_escape" is unable to parse the new debugging location output. [SR-849](https://bugs.swift.org/browse/SR-849) tracks this. Other instructions may have similar problems, but I haven't looked into it any further. The tests pass when run using a single 'sil-opt' invocation.

lattner commented:
Ok.  If I understand you correctly, you're saying that merging this now would cause test suite failures.  If that is the case, we have two options: 1) wait until SR-849 is fixed, or 2) subset out the problematic cases from your test case for now. 

What do you prefer?

ezephir commented:
@lattner SR-849 has been resolved and all the tests pass now. I've rebased again so that the commits appear in the correct order. Everything should be all set for a merge or additional feedback.

Thanks!

jckarter commented:
@swift-ci Please test

jckarter commented:
@adrian-prantl Does this crash look familiar?

```
Assertion failed: (SizeInBits == CI.getTargetInfo().getPointerWidth(0)), function createType, file /Users/buildnode/jenkins/workspace/swift-PR-osx/swift/lib/IRGen/IRGenDebugInfo.cpp, line 1457.
```

adrian-prantl commented:
Yes, I reverted the offending patch pretty much immediately after committing it, you hit a really unlucky time window with the test. Sorry!

-- adrian

> On Mar 9, 2016, at 6:52 PM, Joe Groff <notifications@github.com> wrote:
> 
> @adrian-prantl Does this crash look familiar?
> 
> Assertion failed: (SizeInBits == CI.getTargetInfo().getPointerWidth(0)), function createType, file /Users/buildnode/jenkins/workspace/swift-PR-osx/swift/lib/IRGen/IRGenDebugInfo.cpp, line 1457.
> 
> —
> Reply to this email directly or view it on GitHub.
>

jckarter commented:
Thanks @adrian-prantl . I'll go ahead and merge this then. Thanks @ezephir !

benlangmuir commented:
I reverted this in 0ead929b1286163564fd02507eca38055f1d4bae to try to fix CI.  Example failure: https://ci.swift.org/job/oss-swift-incremental-RA-linux-ubuntu-15_10/3228/consoleFull#1305059968ee1a197b-acac-4b17-83cf-a53b95139a76

benlangmuir commented:
Might be related to @slavapestov's recent changes?  I think this was tested before those went in.
